### PR TITLE
Add quantum metrology tutorial

### PR DESCRIPTION
**Pull request by johannesjmeyer**
**Summary:**
Adds a new tutorial showcasing how variational quantum algorithms can benefit the design of quantum sensing protocols.

**Relevant references:**
[A variational toolbox for quantum multi-parameter estimation](http://arxiv.org/abs/2006.06303)

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

File: demonstrations/tutorial_quantum_metrology.py [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+Ramsay_weights[3] = np.pi / 2
+Ramsay_weights[5] = np.pi / 2
+Ramsay_weights[15] = np.pi / 2
+Ramsay_weights[17] = np.pi / 2
+Ramsay_weights[19] = np.pi / 2
+
+print("Cost for standard Ramsay sensing = {:6.4f}".format(opt_cost(Ramsay_weights)))
+
+##############################################################################
+# We will now make a plot to compare the noise scaling of the above probes.
```
Comment (co9olguy) [inline demonstrations/tutorial_quantum_metrology.py:248]:
```suggestion
# We can now make a plot to compare the noise scaling of the above probes.
```
Just a matter of taste, but I find that "will" is a passive phrasing, and "can" is a more inviting phrasing for the reader

File: demonstrations/tutorial_quantum_metrology.py [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+plt.semilogy(gammas, comparison_costs["standard"], label="Standard")
+plt.xlabel(r"$\gamma$")
+plt.ylabel("Weighted Cramér-Rao bound")
+plt.legend()
+plt.grid()
+plt.show()
+
+##############################################################################
+# We see that after only 20 gradient steps, we already found a sensing protocol
+# that has a better noise resilience than standard Ramsay spectroscopy!
```
Comment (co9olguy) [inline demonstrations/tutorial_quantum_metrology.py:271]: Cool!

File: demonstrations/tutorial_quantum_metrology.py [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+
+
+##############################################################################
+# References
+# ----------
+#
+# 1. Johannes Jakob Meyer, Johannes Borregaard, Jens Eisert.
```
Comment (co9olguy) [inline demonstrations/tutorial_quantum_metrology.py:278]: Would be good to have a relative link in the text itself for this. Check out the Rosalin tutorial for an example on how to do it

File: demonstrations/tutorial_quantum_metrology.py [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+plt.ylabel("Weighted Cramér-Rao bound")
+plt.legend()
+plt.grid()
+plt.show()
+
+##############################################################################
+# We see that after only 20 gradient steps, we already found a sensing protocol
+# that has a better noise resilience than standard Ramsay spectroscopy!
+
+
```
Comment (co9olguy) [inline demonstrations/tutorial_quantum_metrology.py:273]: Felt like it finished just when things were getting exciting :) 

Might be good to have a brief outro? Maybe foreshadow some more advanced things you could do (or that the paper does)?

**Comment by co9olguy** [comment]
One more thought: Might be nice to have a bit more visual flair to the tutorial, maybe you can re-use a figure from your paper?

File: demonstrations/tutorial_quantum_metrology.py [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+# is the Classical Fisher Information Matrix, which we compute using a separate
+# function.
+def CFIM(weights, phi, gamma):
+    p = experiment(weights, phi, gamma=gamma)
+    dp = []
+
+    for idx in range(3):
+        # We use the parameter-shift rule explicitly
+        # to compute the derivatives
+        shift = np.zeros_like(phi)
```
[the remark this answers is not in the document]
Comment (johannesjmeyer) [reply demonstrations/tutorial_quantum_metrology.py:153]: Exactly, but more precisely I need differentiable derivatives. I think this solution is quite fine, internally pennylane isn't doing anything different

File: demonstrations/tutorial_quantum_metrology.py [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+# initialize the weights at random.
+def opt_cost(weights, phi=phi, gamma=gamma, J=J, W=np.eye(2)):
+    return cost(weights, phi, gamma, J, W)
+
+
+# Seed for reproducible results
+np.random.seed(395)
+weights = np.random.uniform(0, 2 * np.pi, NUM_ANSATZ_PARAMETERS + NUM_MEASUREMENT_PARAMETERS)
+
+opt = qml.AdagradOptimizer(stepsize=0.1)
```
[the remark this answers is not in the document]
Comment (johannesjmeyer) [reply demonstrations/tutorial_quantum_metrology.py:216]: Usually I would put a fixed stepsize, but during the numerics of the paper it showed that it's just not really working. In the paper I used a scheme that when the cost increases I reduce the stepsize, but here I tried Adagrad and it worked quite nicely

File: demonstrations/tutorial_quantum_metrology.py [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+.. math:: \operatorname{Cov}(\hat{\boldsymbol{\varphi}}) \geq \frac{1}{n} I^{-1}_{\boldsymbol{\phi}},
+
+where :math:`n` is the number of samples and :math:`I_{\boldsymbol{\phi}}` is the *Classical Fisher Information Matrix*
+with respect to the entries of :math:`\boldsymbol{\phi}`, defined as
+
+.. math:: [I_{\boldsymbol{\phi}}]_{jk} := \sum_l \frac{(\partial_j p_l)(\partial_k p_l)}{p_l},
+
+where we used :math:`\partial_j` as a shorthand notation for :math:`\partial/\partial \phi_j`.
+
+The variational algorithm now proceeds by parametrizing both the probe state :math:`\rho_0 = \rho_0(\boldsymbol{\theta})`
```
[the remark this answers is not in the document]
Comment (johannesjmeyer) [reply demonstrations/tutorial_quantum_metrology.py:47]: I reworked the entire intro, should be better now

File: demonstrations/tutorial_quantum_metrology.py [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+and the POVM :math:`\Pi_l = \Pi_l(\boldsymbol{\mu})`. The parameters :math:`\boldsymbol{\theta}` and :math:`\boldsymbol{\mu}`
```
[the remark this answers is not in the document]
Comment (johannesjmeyer) [reply demonstrations/tutorial_quantum_metrology.py:48]: They don't, but maybe the text is now clearer.

File: demonstrations/tutorial_quantum_metrology.py [hunk 9]
```diff
@@ -0,0 +1,280 @@
...
+##############################################################################
+# As the cost function contains an inversion, we add a small regularization
+# to it to avoid inverting a singular matrix. As additional parameters, we add
+# the weighting matrix `W` and the Jacobian `J`.
+def cost(weights, phi, gamma, J, W, epsilon=1e-10):
+    return np.trace(W @ np.linalg.inv(J.T @ CFIM(weights, phi, gamma) @ J + np.eye(2) * epsilon))
+
+
+##############################################################################
+# To compute the Jacobian, we make use of sympy. Note that we only seek the
```
[the remark this answers is not in the document]
Comment (johannesjmeyer) [reply demonstrations/tutorial_quantum_metrology.py:205]: I know how to do it with sympy. One could surely also do it with autograd, but I actually kind of like sympy so far

**Comment by johannesjmeyer** [comment]
Thanks for you very helpful comments, @co9olguy ! I incorporated them, improved the introduction added a figure and an outro.

File: demonstrations/tutorial_quantum_metrology.py [hunk 10]
```diff
@@ -1,100 +1,139 @@
...
-where :math:`J_{kl} = \partial f_k / \partial \phi_l` is the Jacobian of :math:`\boldsymbol{f}`.
-
-Setup 
------
+where :math:`J_{kl} = \frac{\partial f_k}{\partial \phi_l}` is the Jacobian of :math:`\boldsymbol{f}`.
 
+We now turn to the actual implementation of the scheme.
 """
 import pennylane as qml
 from pennylane import numpy as np
```
Comment (co9olguy) [inline demonstrations/tutorial_quantum_metrology.py]: This part feels a bit orphaned. Would it make sense to move it after the heading directly below?

File: demonstrations/tutorial_quantum_metrology.py [hunk 11]
```diff
@@ -133,19 +174,23 @@ def experiment(weights, phi, gamma=0.0):
...
 
-
 # Make a dry run to be able to draw
-experiment(np.zeros(NUM_ANSATZ_PARAMETERS + NUM_MEASUREMENT_PARAMETERS), np.zeros(3), gamma=0.2)
+experiment(
+    np.zeros(NUM_ANSATZ_PARAMETERS + NUM_MEASUREMENT_PARAMETERS),
+    np.zeros(3),
+    gamma=0.2,
+)
 print(experiment.draw(show_variable_names=True))
```
Comment (co9olguy) [inline demonstrations/tutorial_quantum_metrology.py:183]: Just a thought for future. What's the best way to display long circuits? Would we want them to (optionally) have a "wrapping" after a certain number of characters?

File: demonstrations/tutorial_quantum_metrology.py [hunk 12]
```diff
@@ -269,12 +335,24 @@ def opt_cost(weights, phi=phi, gamma=gamma, J=J, W=np.eye(2)):
 ##############################################################################
 # We see that after only 20 gradient steps, we already found a sensing protocol
 # that has a better noise resilience than standard Ramsay spectroscopy!
+#
+# This tutorial shows that variational methods are useful for quantum metrology.
+# The are numerous avenues open for further research: One could study more intricate
```
Comment (co9olguy) [inline demonstrations/tutorial_quantum_metrology.py:340]:
```suggestion
# The are numerous avenues open for further research: one could study more intricate
```

File: demonstrations/tutorial_quantum_metrology.py [hunk 13]
```diff
@@ -269,12 +335,24 @@ def opt_cost(weights, phi=phi, gamma=gamma, J=J, W=np.eye(2)):
...
+# sensing problems, different noise models and other platforms like optical systems.
+#
+# For more intricate noise models that can't be realized on quantum hardware, the original 
+# work offers a way to move certain parts of the algorithm to the classical side.
+# It also provides extensions of the method to include prior knowledge
+# about the distribution of the underlying parameters or to factor in mutual time dependence
```
Comment (co9olguy) [inline demonstrations/tutorial_quantum_metrology.py:346]:
```suggestion
# about the distribution of the underlying parameters or to factor in mutual time-dependence
```

File: demonstrations/tutorial_quantum_metrology.py [hunk 14]
```diff
@@ -1,100 +1,139 @@
...
-        used to optimize a quantum sensing protocol.
+    :property="og:description": Using a variational algorithm to
+        optimize a quantum sensing protocol.
     :property="og:image": https://pennylane.ai/qml/_images/single_shot.png
 
+*Author: Johannes Jakob Meyer*
+
 In this tutorial we use the variational quantum algorithm from
-`Johannes Jakob Meyer et al. (2020) <https://arxiv.org/abs/2006.06303>`__ to optimize a quantum
+JJM et al. [#meyer2020]_ to optimize a quantum
```
[the remark this answers is not in the document]
Comment (johannesjmeyer) [reply demonstrations/tutorial_quantum_metrology.py:13]: Then I'll remove the author above. Just having Author: Johannes Jakob Meyer and in the next line citing Johannes Jakob Meyer et al. is a bit too much...

File: demonstrations/tutorial_quantum_metrology.py (hunk 11 above)
Comment (johannesjmeyer) [reply demonstrations/tutorial_quantum_metrology.py:183]: I personally like it the way it is with the scrollbar. Imagine you have a wider circuit - then wrapping would make it unreadable...

File: demonstrations/tutorial_quantum_metrology.py [hunk 15]
```diff
@@ -0,0 +1,361 @@
+r"""
+Variationally optimizing measurement protocols
+==================================
```
Comment (co9olguy) [inline demonstrations/tutorial_quantum_metrology.py:3]:
```suggestion
Variationally optimizing measurement protocols
==============================================
```

File: demonstrations/tutorial_quantum_metrology.py [hunk 16]
```diff
@@ -0,0 +1,361 @@
...
+.. figure:: ../demonstrations/quantum_metrology/illustration.png
+    :align: center
+    :width: 50%
+    :target: javascript:void(0)
+
+Here, the encoding process is modeled as a unitary evolution :math:`U(\boldsymbol{\phi})` followed by
+a parameter-independent noise channel :math:`\mathcal{N}`.
+
+Ramsay spectroscopy
+------------------
```
Comment (co9olguy) [inline demonstrations/tutorial_quantum_metrology.py:93]:
```suggestion
-------------------
```

File: demonstrations/tutorial_quantum_metrology.py [hunk 17]
```diff
@@ -0,0 +1,361 @@
...
+    np.zeros(NUM_ANSATZ_PARAMETERS + NUM_MEASUREMENT_PARAMETERS),
+    np.zeros(3),
+    gamma=0.2,
+)
+print(experiment.draw(show_variable_names=True))
+
+
+##############################################################################
+# Evaluating the cost function
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Comment (co9olguy) [inline demonstrations/tutorial_quantum_metrology.py:191]:
```suggestion
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Pull request merged**